pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,168 @@
+					time.Sleep(1 * time.Second)
+				}
+				e.log.Error(err, "failed to update status", "FAR uid", uid)
+				break
+			}
+		}(uid, command)
 	}
+}
 
-	var (
-		stdoutBuf bytes.Buffer
-		stderrBuf bytes.Buffer
-	)
-
-	//containerName := pod.Spec.Containers[0].Name
-	containerName := "manager"
-
-	// Build the Post request for SPDY (shell) executor
-	req := e.clientSet.CoreV1().RESTClient().Post().
-		Resource("pods").
-		Name(pod.Name).
-		Namespace(pod.Namespace).
-		SubResource("exec").
-		Param("container", containerName)
-
-	req.VersionedParams(&corev1.PodExecOptions{
-		Container: containerName,
-		Command:   command,
-		Stdin:     false,
-		Stdout:    true,
-		Stderr:    true,
-		TTY:       false,
-	}, scheme.ParameterCodec)
-
-	execSPDY, err := remotecommand.NewSPDYExecutor(e.config, "POST", req.URL())
-	if err != nil {
-		e.log.Error(err, "failed building SPDY (shell) executor")
-		return "", "", err
+// Exists checks if there is already a running Fence Agent command mapped to the UID
+func (e *Executer) Exists(uid types.UID) bool {
+	e.mapLock.Lock()
+	defer e.mapLock.Unlock()
+	_, exist := e.routines[uid]
+	return exist
+}
+
+// run runs the command in the container and updates the status of the FAR instance maching the UID
+func run(ctx context.Context, logger logr.Logger, uid types.UID, command []string) (stdout, stderr string, err error) {
+	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
+
+	var outBuilder, errBuilder strings.Builder
+	cmd.Stdout = &outBuilder
+	cmd.Stderr = &errBuilder
+
+	for i := 0; i < 5; i++ {
+		err = cmd.Run()
+		if err == nil {
+			break
+		}
+		logger.Error(err, "command failed", "uid", uid, "command", command)
+		time.Sleep(1 * time.Second)
 	}
 
-	// Execute the Post request for SPDY (shell) executor
-	err = execSPDY.Stream(remotecommand.StreamOptions{
-		Stdout: &stdoutBuf,
-		Stderr: &stderrBuf,
-		Tty:    false,
-	})
-	if err != nil {
-		e.log.Error(err, "Failed to run exec command", "stdout", stdoutBuf.String(), "stderr", stderrBuf.String())
+	return outBuilder.String(), errBuilder.String(), err
+}
+
+func (e *Executer) getFARByUID(uid types.UID) (*v1alpha1.FenceAgentsRemediation, error) {
+	farList := &v1alpha1.FenceAgentsRemediationList{}
+	if err := e.List(e.ctx, farList, &client.ListOptions{}); err != nil || len(farList.Items) == 0 {
+		e.log.Error(err, "failed to list FAR", "FAR uid", uid)
+		return nil, err
+	}
+
+	for _, far := range farList.Items {
+		if far.UID == uid {
+			return &far, nil
+		}
+	}
+
+	err := fmt.Errorf("could not find any FAR matching the UID")
+	e.log.Error(err, "failed to get far", "uid", uid)
+
+	return nil, err
+}
+
+func (e *Executer) updateStatus(far *v1alpha1.FenceAgentsRemediation, err error) error {
+	var reason utils.ConditionsChangeReason
+
+	if err == nil {
+		reason = utils.FenceAgentSucceeded
 	} else {
-		e.log.Info("Command has been executed successfully", "stdout", stdoutBuf.String())
+		reason = utils.FenceAgentFailed
```

Comment:
IIUC then after 5 times running the fence agent command, we considered that the cause of failures was not flakiness and we stopped running the fence agent. Right?

Reply:
Correct

Reply:
I assume 5 is just a magic number that makes sense to you. Nothing more than that.
I think there is no harm in executing the command again and again until success. In the meantime, we can log/print the number of failure times or just keep presenting "processing...".

But I would just let the CLI executer run once (and not x times), if it fails return an error in reconciliation and simply try again until success. In the meantime log or present an event for running again the fence agent.
Unless we gain something from running the CLI executer in a for loop except the creation time of the command which should be in favor of looping with for (for x times) instead of looping the reconcile.

Reply:
> I assume 5 is just a magic number that makes sense to you. Nothing more than that.

Of course, since we cannot know what the error means, unless we have a full map telling us which error is temporary and which not, I see the retry just for be completely sure the error is of the first kind.